### PR TITLE
[OSPP][stable-difusion.cpp] achieve new functions

### PR DESCRIPTION
**Pull request by Tenderyi**
In order to implement command line parameters, some functions need to be improved:

- Using ESRGAN to upscale results
- Modify the image size (height, width)
- Output multiple images generated at one time (batch count)

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review agent](https://github.com/flows-network/github-pr-review/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

**Comment by dm4** [comment]
Has this PR been fully implemented? If not, please change the PR status to draft, and if it is complete, please help fix the DCO signature. Thank you.

**Comment by hydai** [comment]
Please fix the lint issues.

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 1]
```diff
@@ -104,6 +108,34 @@ sd_image_t *readControlImage(Span<uint8_t> ControlImage,
   return ControlImg;
 }
 
+void upscalerModel(const char *UpscaleModelPath, uint32_t UpscaleRepeats,
+                  int32_t NThreads, uint32_t Wtype, uint32_t BatchCount, 
+                  sd_image_t *Results) {
+  int upscale_factor = 4;  // unused for RealESRGAN_x4plus_anime_6B.pth
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:114]: Please follow the coding style. We use the LLVM coding style.

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 2]
```diff
@@ -104,6 +108,34 @@ sd_image_t *readControlImage(Span<uint8_t> ControlImage,
...
+  upscaler_ctx_t* upscaler_ctx = new_upscaler_ctx(UpscaleModelPath, NThreads, static_cast<sd_type_t>(Wtype));
+  if (upscaler_ctx == NULL){
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:116]: Use `nullptr`.

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 3]
```diff
@@ -104,6 +108,34 @@ sd_image_t *readControlImage(Span<uint8_t> ControlImage,
   return ControlImg;
 }
 
+void upscalerModel(const char *UpscaleModelPath, uint32_t UpscaleRepeats,
+                  int32_t NThreads, uint32_t Wtype, uint32_t BatchCount, 
+                  sd_image_t *Results) {
+  int upscale_factor = 4;  // unused for RealESRGAN_x4plus_anime_6B.pth
+  upscaler_ctx_t* upscaler_ctx = new_upscaler_ctx(UpscaleModelPath, NThreads, static_cast<sd_type_t>(Wtype));
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:115]: It looks like there is a `new` without a delete. Considering to use `unique_ptr` instead.

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 4]
```diff
@@ -104,6 +108,34 @@ sd_image_t *readControlImage(Span<uint8_t> ControlImage,
...
+  if (upscaler_ctx == NULL){
+    spdlog::error(
+        "[WasmEdge-StableDiffusion] Create upscaler ctx failed."sv);
+  } else {
+    for (uint32_t i = 0; i < BatchCount; i++) {
+        if (Results[i].data == NULL) {
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:121]: `nullptr`.

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 5]
```diff
@@ -104,6 +108,34 @@ sd_image_t *readControlImage(Span<uint8_t> ControlImage,
...
+            continue;
+        }
+        sd_image_t current_image = Results[i];
+        for (uint32_t u = 0; u < UpscaleRepeats; ++u) {
+            sd_image_t upscaled_image = upscale(upscaler_ctx, current_image, upscale_factor);
+            if (upscaled_image.data == NULL) {
+                printf("upscale failed\n");
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:128]: Use `spdlog` instead.

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 6]
```diff
@@ -397,29 +490,68 @@ Expect<uint32_t> SDImageToImage::body(
...
     return static_cast<uint32_t>(ErrNo::RuntimeError);
   }
   std::copy_n(Png, *BytesWritten, OutputBufferSpan.data());
   free(Png);
   free(Results);
-  free(InputImageBuffer);
   free(ControlImageBuffer);
+
   return static_cast<uint32_t>(ErrNo::Success);
+  // int Len;
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:532]: I have no idea why we need these comments.

**Review by hydai — CHANGES_REQUESTED** [review]
Please run `git-clang-format` before submitting the PR.

**Review by hydai — CHANGES_REQUESTED** [review]
Please follow the LLVM naming style.
* `i` -> `I`
* `current_image` -> `CurrentImage`

**Comment by hydai** [comment]
Please run clang-format before you submit the PR.

**Comment by Tenderyi** [comment]
> Please run clang-format before you submit the PR.

I have no idea what's wrong with the format. Cauze I've already run the clang-format-15 to auto check the format. Can you help me with that? Thanks a lot!!!

**Review by hydai — CHANGES_REQUESTED** [review]

**Comment by hydai** [comment]
> > Please run clang-format before you submit the PR.
> 
> I have no idea what's wrong with the format. Cauze I've already run the clang-format-15 to auto check the format. Can you help me with that? Thanks a lot!!!

Well, you just fixed the format in the following commit.

CleanShot 2024-09-23 at 16 05 17

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 7]
```diff
@@ -396,28 +501,53 @@ Expect<uint32_t> SDImageToImage::body(
...
+      free(Results[I].data);
+      Results[I].data = nullptr;
+    }
   }
   *BytesWritten = Len;
   if (OutBufferMaxSize < *BytesWritten) {
     spdlog::error("[WasmEdge-StableDiffusion] Output buffer is not enough."sv);
     free(Png);
     free(Results);
-    free(InputImageBuffer);
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:413]: Why don't we free this variable?

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 8]
```diff
@@ -396,28 +501,53 @@ Expect<uint32_t> SDImageToImage::body(
...
+    upscalerModel(UpscaleModelPath.data(), UpscaleRepeats, NThreads, Wtype,
+                  BatchCount, Results);
+  }
+  // save results
   int Len;
   unsigned char *Png = stbi_write_png_to_mem(
-      reinterpret_cast<const unsigned char *>(Results->data), 0, Results->width,
+      reinterpret_cast<const unsigned char *>(Results), 0, Results->width,
       Results->height, Results->channel, &Len, nullptr);
-  if (OutputPathLen != 0) {
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:404]: Have you verified the Rust SDK will never pass an empty output path? If not, we still need to check the output path is valid.

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 9]
```diff
@@ -396,28 +501,53 @@ Expect<uint32_t> SDImageToImage::body(
...
+  size_t Last = OutputPath.find_last_of(".");
+  std::string DummyName = OutputPath;
+  if (Last != std::string::npos) {
+    DummyName = OutputPath.substr(0, Last);
+  }
+  for (uint32_t I = 0; I < BatchCount; I++) {
+    if (Results[I].data != nullptr) {
+      std::string FinalImagePath;
+      if (I <= 0)
+        FinalImagePath = DummyName + ".png";
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:528]: What if the OutputPath = `ABC.xyz` or `ABC.PNG`? You split the original file name into two parts: name and extension. However, you keep the name but replace the extension with `png` directly.

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 10]
```diff
@@ -374,12 +438,53 @@ Expect<uint32_t> SDImageToImage::body(
...
+          "[WasmEdge-StableDiffusion] The height of image must be greater than 0."sv);
+      free(InputImageBuffer);
+      return static_cast<uint32_t>(ErrNo::InvalidArgument);
+    }
+
+    // Resize input image ...
+    if (Height != static_cast<uint32_t>(ImageHeight) ||
+        Width != static_cast<uint32_t>(ImageWidth)) {
+      uint8_t ResizedHeight = Height;
+      uint8_t ResizedWidth = Width;
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:464]: `Height` and `Width` are `uint32_t` types. Why do we need to reduce the type from `uint32_t` to `uint8_t`?

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 11]
```diff
@@ -374,12 +438,53 @@ Expect<uint32_t> SDImageToImage::body(
...
+      uint8_t *ResizedImageBuffer =
+          (uint8_t *)malloc(ResizedHeight * ResizedWidth * 3);
+      if (ResizedImageBuffer == nullptr) {
+        spdlog::error(
+            "[WasmEdge-StableDiffusion] Allocate memory for resize input image."sv);
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:469]: Should it be `failed to allocate memory ...`?

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 12]
```diff
@@ -374,12 +438,53 @@ Expect<uint32_t> SDImageToImage::body(
...
+      free(InputImageBuffer);
+      InputImageBuffer = ResizedImageBuffer;
+    }
   } else {
     InputImageBuffer =
         stbi_load_from_memory(ImageSpan.data(), ImageSpan.size(), &ImageWidth,
                               &ImageHeight, &Channel, 3);
   }
-  // TODO: Resize image when image size not matches width and height
+  // Resize image when image size not matches width and height
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:487]: This comment should move to the above section.

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 13]
```diff
@@ -396,28 +501,53 @@ Expect<uint32_t> SDImageToImage::body(
...
     free(Png);
     free(Results);
-    free(InputImageBuffer);
     free(ControlImageBuffer);
     return static_cast<uint32_t>(ErrNo::RuntimeError);
   }
   std::copy_n(Png, *BytesWritten, OutputBufferSpan.data());
   free(Png);
   free(Results);
-  free(InputImageBuffer);
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:420]: Ditto.

**Review by hydai — CHANGES_REQUESTED** [review]

File: plugins/wasmedge_stablediffusion/sd_func.cpp (hunk 10 above)
Comment (hydai) [reply plugins/wasmedge_stablediffusion/sd_func.cpp:464]: It's pretty non sense that the type of `InputImageBuffer` will affect the `ResizedHeight` and `ResizedWidth`. The maximum value of `uint8_t` is 255. You truncated the `Height` and `Width` from `uint32_t` into `uint8_t` so the `ResizedImageBuffer` will be extreme smaller than the original image. I don't think this is expected.

Especially, the [`stbir_resize` function](https://github.com/leejet/stable-diffusion.cpp/blob/master/thirdparty/stb_image_resize.h#L341-L347) supports `int` as the input for both `Height` and `Width`.

Comment (Tenderyi) [reply plugins/wasmedge_stablediffusion/sd_func.cpp:464]: I'm sorry for that mistake, and I will change it to `int`.

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 14]
```diff
@@ -272,47 +353,56 @@ Expect<uint32_t> SDTextToImage::body(
...
+    MEM_SPAN_CHECK(UpscaleModelSpan, MemInst, char, UpscaleModelPathPtr,
+                   UpscaleModelPathLen,
+                   "Failed when accessing the Upscaler Image memory."sv)
+    std::string UpscaleModelPath(UpscaleModelSpan.begin(),
+                                 UpscaleModelSpan.end());
+    upscalerModel(UpscaleModelPath.data(), UpscaleRepeats, NThreads, Wtype,
+                  BatchCount, Results);
+  }
+  // Save results
+
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:392]: Redundant empty line.

**Review by hydai — CHANGES_REQUESTED** [review]

**Comment by hydai** [comment]
@grorge123 Could you please take a look at this PR? It seems to change some behavior that you contributed to before.

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 15]
```diff
@@ -272,47 +350,55 @@ Expect<uint32_t> SDTextToImage::body(
...
+  *BytesWritten = saveResults(Results, OutputPath, BatchCount, &Png);
   if (OutBufferMaxSize < *BytesWritten) {
     spdlog::error("[WasmEdge-StableDiffusion] Output buffer is not enough."sv);
+    free(SDCtx);
     free(Png);
     free(Results);
-    free(ControlImageBuffer);
     return static_cast<uint32_t>(ErrNo::RuntimeError);
   }
   std::copy_n(Png, *BytesWritten, OutputBufferSpan.data());
```
Comment (grorge123) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:398]: copy_n is used to write png to OutputBuffer. Is moving it to the saveResults function more proper?

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 16]
```diff
@@ -374,51 +463,96 @@ Expect<uint32_t> SDImageToImage::body(
...
+  *BytesWritten = saveResults(Results, OutputPath, BatchCount, &Png);
   if (OutBufferMaxSize < *BytesWritten) {
     spdlog::error("[WasmEdge-StableDiffusion] Output buffer is not enough."sv);
     free(Png);
     free(Results);
-    free(InputImageBuffer);
-    free(ControlImageBuffer);
     return static_cast<uint32_t>(ErrNo::RuntimeError);
   }
   std::copy_n(Png, *BytesWritten, OutputBufferSpan.data());
```
Comment (grorge123) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:553]: Ditto

**Comment by grorge123** [comment]
Could you solve the CI failing check? It may caused by you changing the CMakeList.

**Comment by hydai** [comment]
The CMake files are broken. It looks like you are trying to apply some weird linter to them.

**Comment by Tenderyi** [comment]
Sorry for misusing clang-format check on CMake file!

**Review by grorge123 — APPROVED** [review]

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 17]
```diff
@@ -215,10 +299,14 @@ Expect<uint32_t> SDCreateContext::body(
...
-  spdlog::info("[WasmEdge-StableDiffusion] Create context."sv);
+  // Check parameters
+  if (ModelPathLen == 0 && diffusionModelPathLen == 0) {
+    spdlog::error("[WasmEdge-StableDiffusion] The following arguments are "
+                  "required: ModelPath / DiffusionModelPath");
+    return static_cast<uint32_t>(ErrNo::InvalidArgument);
+  }
   // Create context and import graph.
-
+  spdlog::info("[WasmEdge-StableDiffusion] Create context."sv);
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:309]: This is redundant info. Maybe it should use `spdlog::debug` instead.

File: plugins/wasmedge_stablediffusion/sd_func.cpp [hunk 18]
```diff
@@ -215,10 +299,14 @@ Expect<uint32_t> SDCreateContext::body(
   if (NThreads == -1) {
     NThreads = get_num_physical_cores();
   }
-
-  spdlog::info("[WasmEdge-StableDiffusion] Create context."sv);
+  // Check parameters
+  if (ModelPathLen == 0 && diffusionModelPathLen == 0) {
+    spdlog::error("[WasmEdge-StableDiffusion] The following arguments are "
+                  "required: ModelPath / DiffusionModelPath");
```
Comment (hydai) [inline plugins/wasmedge_stablediffusion/sd_func.cpp:305]: `sv`.

**Review by hydai — CHANGES_REQUESTED** [review]
There are only two minor issues that block the merge process.

**Comment by Tenderyi** [comment]
> There are only two minor issues that block the merge process.

I’ve addressed the latest issues and the PR is now ready for merging. When you have a moment, could you please review and approve it?

**Review by dm4 — APPROVED** [review]
Since @hydai currently has some matters to attend to, after discussing with him, he also agrees that we can approve and merge this pull request.

**Pull request merged**